include/circt/Support/PrettyPrinterHelpers.h

```diff
@@ -166,6 +166,55 @@ class TokenStringSaver : public PrettyPrinter::Listener {
   void clear() override;
 };
 
+template <typename Callable, typename Data>
+class PrintEventListener : public PrettyPrinter::Listener {
+
+  /// List of all the unique data associated with each callback token.
+  /// Note: Doesn't own Data !
+  SmallVector<Data *> list;
+  /// The storage for the callback, as a function object.
+  std::function<void(uint32_t)> callbackObj;
+
+  /// Note: Callable class must implement a callable with signature:
+  /// void (Data*)
+  /// Create and initialize a function object, capturing the callable object and
+  /// a pointer to the data vector. The callable object contains all the values
+  /// shared across the callback tokens. The Data list indexed by the parameter
+  /// provides the unique data for each callback.
+  void init(Callable &c) {
+    // Ensure init is called only once.
+    assert(!callbackObj);
+    assert(list.empty());
+    // May result in heap allocation. Pointer to the callback object and data
+    // vector is captured by value.
+    callbackObj =
```

Comment:
Can drop callbackObj and just stash `c` now.

Then the below `print` can just the current body of this lambda?